### PR TITLE
Add support for heterogeneous compute

### PR DESCRIPTION
**Pull request by jcranmer-intel**
This is an incomplete pull request at the moment, but I feel it would be better to continue the conversation with some actual code behind it. This is to fix issue #2244 .

At the moment, I've tested this with a basic support for `clang -fopenmp` offload runs. Specifically, the example code I used would be the example here: https://godbolt.org/z/rvGeqc.

Starting with the UX: I implemented the device viewer pane using a slightly-modified copy of the IR viewer. It is insufficient at this moment; I haven't added support for device selection or device-specific arguments. Some of my testing identifies other issues: the same asm filters are applied to both the regular ASM view and the device ASM view, which may not be the best UI. In general, UI is not my strong suit, and I'd personally prefer to focus my efforts on backend implementation issues than UI issues if someone else is willing to pick up the slack.

The backend only adds support for the `clang -fopenmp` heterogeneous flow, which is probably the simplest flow as you actually get simple assembly files of both host and offload neatly packaged into one. GCC's OpenMP implementation packs it as section data; the SYCL offload embeds a .bc file instead. My plan to support all these flows is to require the use of a device extractor tool (deviceExtractor in the properties files), and this tool would have to live in the tools repository. I've prototyped it as a python script that extracts files into a directory as well as outputting a JSON blob indicating which generated file is the host and which generated outputs correspond to which devices.

The first commit in this PR should have all of the relevant information for understanding the backend portion of the code (except the device extractor code itself), while the second commit is all the necessary changes for a rudimentary UI portion.

Feedback would be greatly appreciated.

### CODE REVIEW

File: lib/compilers/clang.js
Comment (apmorton) [inline lib/compilers/clang.js:52]: nit: can we reverse these conditions to avoid executing the regex on compilers that don't claim support?

File: lib/compilers/clang.js
Comment (apmorton) [inline lib/compilers/clang.js:67]: stray console log

File: lib/compilers/clang.js
Comment (apmorton) [inline lib/compilers/clang.js:60]: Generally we wouldn't throw in this case, you would modify `result` to contain the error information in some way and then return early. This way the error information flows to the user on the client side.

**Review by apmorton — COMMENTED** [review]
The backend code is good other than the few comments I made.

UI is not my strong suit either, but things largely look to be in the right place.

File: etc/config/c++.defaults.properties
Comment (mattgodbolt) [inline etc/config/c++.defaults.properties:28]: Whoops: thanks for fixing this. I run with a local properties file that overrides this so hadn't spotted it!

File: lib/compilers/clang.js
Comment (mattgodbolt) [inline lib/compilers/clang.js:56]: Loving the use of `const` everywhere: thank you!

File: lib/compilers/clang.js
Comment (mattgodbolt) [inline lib/compilers/clang.js:69]: Really nice, thanks

**Review by mattgodbolt — COMMENTED** [review]
Looking amazing so far. Have paused while I try to run locally to test: how might I get a 'deviceExtractor' to test against?

File: static/explorer.scss
Comment (mattgodbolt) [inline static/explorer.scss:95]: From my experimentation, and assuming we don't add too many other tool icons on the view, I think we should make this larger, maybe 18 or 20?

File: static/panes/compiler.js
Comment (mattgodbolt) [inline static/panes/compiler.js:946]: I think we can consider removing this tracking stuff: we should discuss whether we want to support multiple open device views.

File: static/panes/compiler.js
Comment (mattgodbolt) [inline static/panes/compiler.js:947]: If we're keeping this logic, can you initialise this to `false` in the constructor please?

File: static/panes/device-view.js
Comment (mattgodbolt) [inline static/panes/device-view.js:27]: Thanks for doing all this. You can see all the copy/paste...one day we'll clean this up...

**Review by mattgodbolt — APPROVED** [review]
This is fantastic! If you're happy I'll merge and we can start iterating on minor fixes. We discussed in Slack about maybe javacsripting the device extractor; but you correctly point out we should probably wait to see how the other extractors go :)

**Comment by mattgodbolt** [comment]
Any thoughts on the comments I made @jcranmer-intel ? Super minor but I'll merge once they're in :+1:

**Comment by mattgodbolt** [comment]
Ping on this @jcranmer-intel  :)

**Comment by partouf** [comment]
I stayed out of this because it's all greek to me, but I fear I have to ask some questions:
* What's the status on this?
* What can we do?
* And how can we test this?
* Can we document a bit of this in layman terms?

**Comment by partouf** [comment]
> I stayed out of this because it's all greek to me, but I fear I have to ask some questions:
> 
> * What's the status on this?
> * What can we do?
> * And how can we test this?
> * Can we document a bit of this in layman terms?

@mattgodbolt @jcranmer-intel code Any answers to any of these questions?

**Comment by mattgodbolt** [comment]
Excellent question @partouf ! I'm liking where this is going but I'm waiting on @jcranmer-intel - I had only minor comments but before I merge a feature like this I'd like to know we have the support and backing of the original author :)

**Comment by mattgodbolt** [comment]
Ping on this :) Would be nice to merge. @jcranmer-intel

**Comment by mattgodbolt** [comment]
Apologies, my attempt to follow the instrictuions on GH failed...will force push shortly and fix.

**Comment by mattgodbolt** [comment]
Closed in favour of #2836

**Pull request closed without merging**
